[LLVM][SelectionDAG] Align poison/undef binop folds with IR.

paulwalker-arm opened the pull request:
The "at construction" binop folds in SelectionDAG::getNode() has different behaviour when compared to the equivalent LLVM IR. This PR makes the behaviour consistent while also extending the coverage to include signed/unsigned max/min operations.

https://godbolt.org/z/nKG5d7rsG shows the IR equivalent for min/max.

paulwalker-arm committed:
[LLVM][SelectionDAG] Align poison/undef binop folds with IR.

The "at construction" binop folds in SelectionDAG::getNode() has
different behaviour when compared to the equivalent LLVM IR. This PR
makes the behaviour consistent while also extending the coverage to
include signed/unsigned max/min operations.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7895,16 +7904,30 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
                                            : getAllOnesConstant(DL, VT);
+    case ISD::SMAX:
+      return N2.getOpcode() == ISD::POISON
+                 ? getPOISON(VT)
```
Can reuse N2 instead of constructing a new poison

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7875,13 +7875,22 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+      case ISD::SMIN:
+      case ISD::UMAX:
+      case ISD::UMIN:
+        return N2.getOpcode() == ISD::POISON ? getPOISON(VT) : getUNDEF(VT);
```
This is equivalent to just returning N2?

paulwalker-arm committed:
Add unittest to improve test coverage for SelectionDAG::getNode().

It is easier to validate the undef/poison behaviour by probing getNode
directly rather than relying on llc tests.

paulwalker-arm commented:
I think I've uncovered incorrect results for several cases, at least when compared to the results generated for the equivalent IR. There's a new SelectionDAG unit test to show this and I corrected some of them as a side effect of adding support for min/max.  If you agree these are genuine issues then I'm happy to fix the rest as part of this PR or a followup. Please let me know either way.

arsenm commented on llvm/unittests/CodeGen/SelectionDAGNodeConstructionTest.cpp:
```diff
@@ -0,0 +1,454 @@
+    SMDiagnostic SMError;
+    M = parseAssemblyString(Assembly, SMError, Context);
+    if (!M)
+      report_fatal_error(SMError.getMessage());
```
report_fatal_error is deprecated. Either use reportFatalUsageError, though in this context maybe the gtest ASSERT_*s would be more appropriate

paulwalker-arm replied:
I've converted them to ASSERT_TRUE.

arsenm commented on llvm/unittests/CodeGen/SelectionDAGNodeConstructionTest.cpp:
```diff
@@ -0,0 +1,454 @@
+
+    DAG = std::make_unique<SelectionDAG>(*TM, CodeGenOptLevel::None);
+    if (!DAG)
+      report_fatal_error("DAG?");
```
Messages should be more descriptive

arsenm commented on llvm/unittests/CodeGen/SelectionDAGNodeConstructionTest.cpp:
```diff
@@ -0,0 +1,454 @@
+    InitializeAllTargetMCs();
+  }
+
+  void SetUp() override {
```
Probably should have a common DAG unit test base for all of this boilerplate

arsenm commented on llvm/unittests/CodeGen/SelectionDAGNodeConstructionTest.cpp:
```diff
@@ -0,0 +1,454 @@
+  EXPECT_EQ(DAG->getNode(ISD::ADD, DL, Int32VT, Undef, Undef), Undef);
+}
+
+TEST_F(SelectionDAGnodeConstructionTest, AND) {
```
Capitalize Node throughout

arsenm commented on llvm/unittests/CodeGen/SelectionDAGNodeConstructionTest.cpp:
```diff
@@ -0,0 +1,454 @@
+
+TEST_F(SelectionDAGnodeConstructionTest, SADDSAT) {
+  SDLoc DL;
+  auto Int32VT = EVT::getIntegerVT(Context, 32);
```
No auto

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7869,21 +7870,29 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
-    switch (Opcode) {
-    case ISD::XOR:
-      if (N1.isUndef())
+    if (N1.isUndef()) {
```
This should consider poison separately from undef?

arsenm reviewed (approved):
LGTM with test nit. Don't use auto for types and you don't need to use the get* APIs for trivial types

paulwalker-arm committed:
Use MVT::i32 instead of Int32VT indirection.
